cli/src/commands/workspace.rs

```diff
@@ -306,22 +360,36 @@ fn cmd_workspace_update_stale(
     // operation, then merge the concurrent operations. The wc_commit_id of the
     // merged repo wouldn't change because the old one wins, but it's probably
     // fine if we picked the new wc_commit_id.
-    let known_wc_commit = {
-        let mut workspace_command = for_stale_working_copy(command, ui)?;
+    let (known_wc_commit, recovered) = {
+        let (mut workspace_command, recovered) = for_stale_working_copy(command, ui)?;
         workspace_command.maybe_snapshot(ui)?;
+
         let wc_commit_id = workspace_command.get_wc_commit_id().unwrap();
-        workspace_command.repo().store().get_commit(wc_commit_id)?
+        (
+            workspace_command.repo().store().get_commit(wc_commit_id)?,
+            recovered,
+        )
     };
     let mut workspace_command = command.workspace_helper_no_snapshot(ui)?;
 
     let repo = workspace_command.repo().clone();
     let (mut locked_ws, desired_wc_commit) =
         workspace_command.unchecked_start_working_copy_mutation()?;
     match check_stale_working_copy(locked_ws.locked_wc(), &desired_wc_commit, &repo)? {
-        WorkingCopyFreshness::Fresh | WorkingCopyFreshness::Updated(_) => writeln!(
-            ui.stderr(),
-            "Nothing to do (the working copy is not stale)."
-        )?,
+        WorkingCopyFreshness::Fresh | WorkingCopyFreshness::Updated(_) => {
+            if !recovered {
+                writeln!(
+                    ui.stderr(),
+                    "Nothing to do (the working copy is not stale)."
+                )?;
+            }
+        }
+        WorkingCopyFreshness::MissingOperation =>
+        // We should have already recovered from this situation above, so there must
+        // have been a concurrent operation.
+        {
+            return Err(user_error("Concurrent working copy operation. Try again."))
```

Comment:
Another idea is to not map `OpStoreError` to `MissingOperation` by `check_stale_working_copy()`, and raise the original error here.

Then, `snapshot_working_copy()` attaches the hint if `check_stale_working_copy()` returned an Err.

Reply:
Because the error here means another weird thing happened?

I don't have strong opinion, but the underlying error might provide more context why the new operation got lost.

Reply:
Ah, I see. I was thinking that this is a similar case to WorkingCopyStale/SiblingOperation below, which doesn't report the underlying error but just says "Try again" in a weird state. I'll let @martinvonz make the call here.

Reply:
I checked with @martinvonz and I'll go with what @yuja says.

Reply:
I tried doing this, but this would require making WorkingCopyFreshness carry an error, which I think it can't (it derives PartialEq and errors are not PartialEq). If @yuja doesn't feel strongly about this, then hopefully the existing code is fine.

Reply:
> would require making WorkingCopyFreshness carry an error, which I think it can't (it derives PartialEq and errors are not PartialEq)

PartialEq is there only to make tests convenient. We have dropped it from errors before. The workaround is to use `matches!(err, MyError::MyVariant(_))` or similar.

(I haven't caught up with the rest of the thread yet)

Reply:
Having read through this, it sounds like Yuya's suggestion might simplify a little, so I think it's worth attempting it. If it gets complicated for some reason, feel free to leave it as is IMO (Yuya also seems fine with that).

Reply:
Since `check_stale_working_copy()` returns `Result<_, OpStoreError>`, maybe we can check if the error matches `ObjectNotFound` or not. It can also mean the ancestor operation was missing, but we'll probably want to do `update-stale` in that case.

Reply:
Thanks @yuja for the detailed instructions - I think I've done what you've suggested (making `check_stale_working_copy()` return the missing-operation case in the Err OpStoreError instead of Ok MissingOperation), and as @martinvonz says, it indeed makes things simpler.